src/main/kotlin/com/openlattice/data/DataGraphService.kt

```diff
@@ -559,39 +586,61 @@ open class DataGraphService(
         return null
     }
 
+    /**
+     * Checks entity types of associations against the allowed src and dst entity types in association type.
+     * @param srcAssociationEntitySetIds The entity set ids of the src entities in associations mapped by their association entity set id.
+     * @param dstAssociationEntitySetIds The entity set ids of the dst entities in associations mapped by their association entity set id.
+     * @param bidirectionalEntitySetIds The src and dst entity set id pairs of bidirectional associations mapped by the association entity set id.
+     * @param associationTypes Map of association entity set ids and their association types.
+     */
     private fun checkAssociationEntityTypes(
-            srcAssociationEntitySetIds: Map<UUID, Set<UUID>>, dstAssociationEntitySetIds: Map<UUID, Set<UUID>>
+            srcAssociationEntitySetIds: Map<UUID, Set<UUID>>,
+            dstAssociationEntitySetIds: Map<UUID, Set<UUID>>,
+            bidirectionalEntitySetIds: Map<UUID, Set<Pair<UUID, UUID>>>, // edge - (src, dst)
+            associationTypes: Map<UUID, AssociationType> // entityset id - associationtype
```

Comment:
It's probably also better to load all association types + entity types in a few bulk calls here instead of having `integrateAssociations` handle part of it while iterating and having `checkAllowedEntityTypesOfAssociation` handle the rest:
1) load all the entity type ids from the combined keySets and values of `srcAssociationEntitySetIds` and `dstAssociationEntitySetIds`
2) Load all the association types in bulk from the entity type ids of the association entity sets
3) Load all the src+dst entity types in bulk from the entity type ids of the src+dst entity sets

You can also avoid iterating over bidirectional and non-bidirectional (unidirectional? idk) association types separately / writing separate functions to check their allowed entity types. Bidirectional is basically shorthand for saying that `src == src + dst` and `dst == src + dst`, so you can just combine their src and dst entityTypeId sets and use that combined set for both src and dst. i.e. something like

```
val entityTypeIdsByEntitySetId = ...
val associationTypesByAssociationTypeId = ...

(srcAssociationEntitySetIds.keys + dstAssociationEntitySetIds.keys).forEach { edgeEntitySetId -> 
  val associationTypeId = entityTypeIdsByEntitySetId.getValue(edgeEntitySetId)
  val associationType = associationTypesByAssociationTypeId.getValue(associationTypeId)

  val requestedSrcEntityTypeIds = srcAssociationEntitySetIds.getValue(edgeEntitySetId).map{ entityTypeIdsByEntitySetId.getValue(it)
  val requestedDstEntityTypeIds = dstAssociationEntitySetIds.getValue(edgeEntitySetId).map{ entityTypeIdsByEntitySetId.getValue(it)

  val allowedSrcEntityTypeIds = associationType.src
  val allowedDstEntityTypeIds = associationType.dst

  if (associationType.isBidirectional) {
    allowedSrcEntityTypeIds.addAll(associationType.dst)
    allowedDstEntityTypeIds.addAll(associationType.src)
  }

  if (!allowedSrcEntityTypeIds.containsAll(requestedSrcEntityTypeIds) || !allowedDstEntityTypeIds.containsAll(requestedDstEntityTypeIds)) {
    // much sadness
  }

}
```

^btw no guarantees on the compilability of my kotlin

Reply:
"Bidirectional is basically shorthand for saying that src == src + dst and dst == src + dst": I don't think, that is enough to check, because in this case you could still mix and match. For example, allowed src types are `A, B` and allowed dst types are `C,D`, that check would allow `src = A` and `dst = B`

Reply:
I think that might still be correct, although that's a good point and isn't well-defined anywhere.